admin/Default/1.6/universe_create_locations.php

```diff
@@ -1,66 +1,80 @@
 <?php
 require_once(get_file_loc('SmrGalaxy.class.inc'));
 
-//universe_create_locations.php
-$container = $var;
-$container['url'] = '1.6/universe_create_save_processing.php';
-$container['body'] = '1.6/universe_create_sectors.php';
-
-$PHP_OUTPUT.= create_echo_form($container);
-//get totals
-//$totalLocs[5] = 0;
-//$totalLocs[6] = 0;
 $locations =& SmrLocation::getAllLocations();
 
+// Initialize all location counts to zero
+$totalLocs = array();
+foreach ($locations as &$location) {
+	$totalLocs[$location->getTypeID()] = 0;
+}
+
+// Determine the current amount of each location
 $galSectors =& SmrSector::getGalaxySectors($var['game_id'],$var['gal_on']);
 foreach ($galSectors as &$sector) {
 	$sectorLocations =& $sector->getLocations();
 	foreach ($sectorLocations as &$sectorLocation) {
 		$totalLocs[$sectorLocation->getTypeID()]++;
 	} unset($sectorLocation);
 } unset($sector);
+$template->assignByRef('TotalLocs', $totalLocs);
+
 $galaxy =& SmrGalaxy::getGalaxy($var['game_id'],$var['gal_on']);
-$PHP_OUTPUT.= 'Working on Galaxy : ' . $galaxy->getName() . ' (' . $galaxy->getGalaxyID() . ')<br />';
-$PHP_OUTPUT.= '<table class="standard">';
+$template->assignByRef('Galaxy', $galaxy);
 
+// Set any extra information to be displayed with each location
+$locText = array();
+$locTypes = array();
 foreach ($locations as &$location) {
-//	if (isset($loc_array['Do Not List']) && $loc_array['Do Not List']) continue;
 	$extra = '<span class="small"><br />';
 	if ($location->isWeaponSold()) {
-		//$extra = '<table class="nobord right">';
+		$locTypes['Weapons'][] = $location->getTypeID();
 		$weaponsSold =& $location->getWeaponsSold();
 		foreach($weaponsSold as &$weapon) {
 			$extra .= $weapon->getName() . '&nbsp;&nbsp;&nbsp;(' . $weapon->getShieldDamage() . '/' . $weapon->getArmourDamage() . '/' . $weapon->getBaseAccuracy() . ')<br />';
 		} unset($weapon);
-			//$extra .= '<tr><td class="right"><span class="small">' . $WEAPONS[$wep_id]['Weapon Name'] . '</span></td><td class="left"><span class="small">(' . $WEAPONS[$wep_id]['Shield Damage'] . '/' . $WEAPONS[$wep_id]['Armor Damage'] . '/' . $WEAPONS[$wep_id]['Accuracy'] . ')</span></td></tr>';
-		//$extra .= '</table>';
 	}
-	if ($location->isShipSold()) {
+	elseif ($location->isShipSold()) {
```

Comment:
This change means that any shop that can sell multiple things will only show one type of thing that it can sell - is it possible to avoid that?

Reply:
If a shop sells items from more than one category, then the whole categorization scheme breaks down. Though it is not an enforced rule, the game fortunately follows it.

If for some reason we decide to add a location that violates that rule, then what do we do? Do we display it in only one of the categories (and have people unable to find it if they are looking for it in another category)? Do we display it in multiple categories and then implement some logic to make sure it's not double-counted? It's a tricky situation once this rule is violated...

Given the complexity of the game and the rate of code development, it seems reasonable to code for the game we have rather than the game we _could_ have in the future (but only if it makes the current game much more manageable, which in the case of this list, I bet anyone who has created a game would agree it does).

If you decide to go with this as is, I'd be happy to add a comment in the code explicitly pointing out this restriction. Then if someone decides to violate it in the future, they'll see this and need to think about how to make this list manageable under the new assumptions.

Reply:
Well the thing is that there's an admin tool right now that lets admins edit locations, so it is already in the game (albeit whether it's actually used is another matter) and this tools supports seeing it.  I think the easiest solution for the categorisation is just to categorise as the first type of good in the list, and then add the total list later, nicest way I can think to do that is something like:
```php
function addCategory(&$added, &$arr, $typeID) {
    if (!$added) {
        $arr[] = $typeID;
    }
    $added = true;
}

foreach ($locations as &$location) {
	$categoryAdded = false
	if ($location->isWeaponSold()) {
		addCategory(categoryAdded, $locTypes['Weapons'], $location->getTypeID())
		$locTypes['Weapons'][] = $location->getTypeID();
		...